---
title: Getting Started with the NIAGADS-OA API
description: This guide explains how to construct and execute API requests and interpret responses.
---

The NIAGADS-OA API establishes a simple framework for sharing and integration of NIAGADS datasets
and annotated gene and variant records into analysis pipelines. Use of the API requires either a
(command line) tool or requests library that supports the HTTP protocol to transfer data,
permitting large-scale programmatic access independent of any specific platform or programming language.

<Note title="Contribute">
	Interested in helping us improve? Check out our [GitHub
	repository](https://github.com/NIAGADS/niagads-api) to get started! From
	feature suggestions to code-wrappers to bug reports, all contributions are
	welcome.
</Note>

## Tools and Libraries

Third-party tools can be used for communicating with the API and for exploring JSON responses. These include:

| **Tool**                                                      | **Type**                                     |
| ------------------------------------------------------------- | -------------------------------------------- |
| [curl](https://curl.se/)                                      | command line                                 |
| [HTTPie](https://httpie.io/)                                  | command line // desktop (Windows) // web app |
| [Google Chrome](https://www.google.com/chrome/)               | web browser                                  |
| [Mozilla Firefox](https://www.mozilla.org/en-US/firefox/new/) | web browser                                  |
| [Safari](https://www.apple.com/safari/)                       | web browser (Mac)                            |

As well as any programming language package that supports the HTTP protocol, such as:

| **Package**                                                                                                          | **Language** |
| -------------------------------------------------------------------------------------------------------------------- | ------------ |
| [requests](https://pypi.org/project/requests/)                                                                       | Python       |
| [aiohttp](https://pypi.org/project/aiohttp/)\*                                                                       | Python       |
| [request](https://cran.r-project.org/web/packages/request/index.html)                                                | R            |
| [httr2](https://httr2.r-lib.org/)\*                                                                                  | R            |
| [Faraday](https://github.com/lostisland/faraday)                                                                     | ruby         |
| [net/http](https://pkg.go.dev/net/http)                                                                              | go           |
| [java.net.http](https://docs.oracle.com/en/java/javase/21/docs/api/java.net.http/java/net/http/package-summary.html) | Java         |
| [Fetch API](https://www.w3schools.com/jsref/api_fetch.asp)                                                           | Javascript   |
| [HTTP::Tiny](https://perldoc.perl.org/HTTP::Tiny)                                                                    | Perl         |
| [LWP::UserAgent](https://metacpan.org/release/GAAS/libwww-perl-5.69/view/lib/LWP/UserAgent.pm)                       | Perl         |

\*recommended

## API Endpoints

Communicating with the NIAGADS Open Access API involves making calls to API endpoints. Each endpoint
represents both the resource being queried and specific API functionality.

### Request URL

**The HTTP URL that corresponds to the latest version of an API endpoint
is:**<br/> `https://api.niagads.org/<resource>/<operation>`

In the request URL, `<resource>` is the name of the NIAGADS knowledge base and `<operation>` is
the type of request being made to the API.

<Note title="Other Versions">
    The HTTP URL of an endpoint corresponding to a specific major version of the
    API is `https://api.niagads.org/v<version>/<resource>/<operation>`,
    where `<version>` is the NIAGADS-OA API major version.
</Note>

### Resources

Allowable values for `<resource>` are summarized below:

| **Resource** | **Knoweldge Base**                      |
| ------------ | --------------------------------------- |
| genomics\*\* | GenomicsDB                              |
| filer        | FILER                                   |
| advp\*       | ADVP                                    |
| all\*        | search across all Open Access Resources |

\* coming soon<br/>
\*\* under construction

### Operations

Operations define particular functions or services provided by the API.  
Allowable values for `<operation>` are as follows:

| **Operation** | **Type**                  | **Description**                                                              |
| ------------- | ------------------------- | ---------------------------------------------------------------------------- |
| status        | Status                    | Get API description and versioning information                               |
| record        | Data Retrieval            | Retrieve entity metadata, data, or annotations by ID                         |
| search        | Search                    | Find entities matching search criteria                                       |
| data          | Search and Data Retrieval | Retrieve data or annotations for one or more entity matching search criteria |
| service       | Data Retrieval            | Retrieve data formatted for input to NIAGADS visualization tools or services |
| dictionary    | Search and Data Retrieval | Search or retrieve controlled vocabularies and ontologies                    |

### Entities

`record` operations are further qualified by the `entity` type being queried.

All entities are assigned a unique **stable** identifier (ID). These are consistent across all NIAGADS Open Access
resources and are stored in a record's `id` property.

**The HTTP URL for retrieving the record for a specific entity is:** <br/>
`https://api.niagads.org/<resource>/record/<entity>/<id>`

The NIAGADS Open Access API supports queries for the following entities:

| **Entity** | **Record Type**                              | **Description**            | **Identifier**                                                             | **Resource**    |
| ---------- | -------------------------------------------- | -------------------------- | -------------------------------------------------------------------------- | --------------- |
| gene       | annotations, genetic associations            | Ensembl genes              | [Ensembl ID](https://useast.ensembl.org/info/genome/stable_ids/index.html) | genomics, advp  |
| variant    | annotations, genetic associations            | SNVs, short INDELS         | `chr:pos:ref:alt`                                                          | genomics, advp  |
| span\*     | overlapping genomic features, data summaries | arbitrary genomic regions  | `chr:pos:start-end`                                                        | genomics        |
| track      | metadata                                     | data tracks                | `NG[unique alpha-numeric code]`                                            | genomics, filer |
| collection | metadata                                     | set of related data tracks | human-readable string                                                      | genomics, filer |

\* coming soon <br/>

<Note title="Alternative Identifiers">
	The NIAGADS-OA API will also accept industry standard alternative
	identifiers, such as refSNP ID for variants, and official symbols, and
	Entrez (NCBI Gene) IDs for genes. **However, these alternatives may not map
	unambiguously to NIAGADS-OA entities.**
</Note>

## Rate Limits

<Note title="Coming soon">
	More information about rate limits on API queries coming soon.
</Note>

## API Responses

The NIAGADS Open Access API uses JSON as its standard response format, although some endpoints can also return
plain text results (formatted as tab-delimited TEXT, YAML, BED, VCF) by setting a `format` parameter.

<Note title="Want XML?">
	Make a feature request on
	[GitHub](https://github.com/NIAGADS/niagads-api/discussions)!
</Note>

### JSON Response Format

The following fields are included in all successful, non-`service` API responses.

```json
"data": [],
"request": {
    "request_id": "string",
    "endpoint": "string",
    "parameters": {
        "param1": "any of: string"
    }
},
"pagination": {
    "page": "integer",
    "total_num_pages": "integer",
    "paged_number_of_records": "integer",
    "total_num_records": "integer"
}
```

**`data`**: a list of one or more records or data points; format of list entries
will vary by the resource and type of record or data being queried by the endpoint

**`request`**: details about the originating request that generated the response

**`pagination`**: pagination information, if results are paged (_Optional_)

### Status Codes

API responses include a HTTP status code indicating whether the request was successfully completed.

| **Status Code**               | **Type**     | **Description**                                                    |
| ----------------------------- | ------------ | ------------------------------------------------------------------ |
| **200 OK**                    | Success      | successful response                                                |
| **404 Not Found**             | Client Error | record not found                                                   |
| **422 Unprocessable Content** | Client Error | parameter validation error                                         |
| **429 Too Many Requests**     | Client Error | too many requests in a given amount of time, exceeding rate limits |
| **500 Internal Server Error** | Server Error | system or runtime error                                            |
